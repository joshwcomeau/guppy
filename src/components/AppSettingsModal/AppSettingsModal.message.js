// @flow
import { defineMessages } from 'react-intl';

const messages = {
  main: {
    title: {
      id: 'appSettingsModal.main.title',
      defaultMessage: 'Preferences',
      description: 'Title for the whole "Preferences" part',
    },
    errorMessage: {
      id: 'appSettingsModal.main.errorMessage',
      defaultMessage: "Path doesn't exist.",
      description: "Error message if project folder doesn't exist",
    },
    errorPrompt: {
      id: 'appSettingsModal.main.errorPrompt',
      defaultMessage:
        'Please check your default project path or use the directory picker to select the path.',
      description: "Prompt message if project folder doesn't exist",
    },
    save: {
      id: 'appSettingsModal.main.save',
      defaultMessage: 'Save',
      description: 'Text for the save button',
    },
  },
  general: {
    title: {
      id: 'appSettingsModal.general.title',
      defaultMessage: 'General',
      description: 'Title for the "General" part',
    },
    projectPath: {
      id: 'appSettingsModal.general.projectPath',
      defaultMessage: 'Default Project Path',
      description: 'Title for "Default Project Path" under the "General" part',
    },
    projectType: {
      id: 'appSettingsModal.general.projectType',
      defaultMessage: 'Default Project Type',
      description: 'Title for "Default Project Type" under the "General" part',
    },
  },
  privacy: {
    title: {
      id: 'appSettingsModal.privacy.title',
      defaultMessage: 'Privacy',
      description: 'Title for the "Privacy" part',
    },
    enableUsageTracking: {
      id: 'appSettingsModal.privacy.enableUsageTracking',
      defaultMessage: 'Enable anonymous usage tracking',
      description:
        'Title for "Enable anonymous usage tracking" under the "privacy" part',
    },
  },
};

const definedMessages = defineMessages(messages);

export default definedMessages;
