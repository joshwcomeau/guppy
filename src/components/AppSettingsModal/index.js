// @flow
export { default } from './AppSettingsModal';
