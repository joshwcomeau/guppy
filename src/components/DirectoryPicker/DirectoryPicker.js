// @flow
import { remote } from 'electron';
import React, { PureComponent } from 'react';
import styled from 'styled-components';
import IconBase from 'react-icons-kit';
import { folderPlus } from 'react-icons-kit/feather/folderPlus';

import { COLORS } from '../../constants';

import TextButton from '../TextButton';
import TextInput from '../TextInput';

type Props = {
  path: string,
  pathToSelectMessage: ?string,
  inputEditable: ?boolean,
  isFocused: ?boolean,
  onFocus: string => void,
  onSelect: string => void,
};

class DirectoryPicker extends PureComponent<Props> {
  static defaultProps = {
    pathToSelectMessage: 'Select the directory',
  };

  updatePath = (ev: SyntheticEvent<*>) => {
    ev.preventDefault();
    remote.dialog.showOpenDialog(
      {
        message: this.props.pathToSelectMessage,
        properties: ['openDirectory'],
      },
      (paths: ?[string]) => {
        // The user might cancel out without selecting a directory.
        // In that case, do nothing.
        if (!paths) {
          return;
        }

        // Only a single path should be selected
        const [firstPath] = paths;
        this.props.onSelect(firstPath);
      }
    );
  };

  getFullProjectPath(path: string) {
    return path.replace(/^\//, '');
  }

  render() {
    const { path, inputEditable, onFocus, onSelect, isFocused } = this.props;

    // Join the projectHome with the prospective project ID
    // Hide the leading forward-slash, on Mac/Linux
    const fullProjectPath = this.getFullProjectPath(path);

    return (
      <Wrapper>
        {!inputEditable ? (
          <DirectoryButton onClick={this.updatePath} hideCursor={true}>
            {fullProjectPath}
          </DirectoryButton>
        ) : (
          <TextInput
            onFocus={onFocus}
            isFocused={isFocused}
            value={fullProjectPath}
            onChange={ev => onSelect(ev.target.value)}
            autoFocus
          >
            <ButtonPositionAdjuster>
              <IconWrapper onClick={this.updatePath}>
                <IconBase size={22} icon={folderPlus} />
              </IconWrapper>
            </ButtonPositionAdjuster>
          </TextInput>
        )}
      </Wrapper>
    );
  }
}

const Wrapper = styled.div`
  color: ${COLORS.gray[400]};
`;

const ButtonPositionAdjuster = styled.div`
  transform: translateY(-2px);
`;

const DirectoryButton = styled(TextButton)`
  color: ${COLORS.gray[600]};
  text-decoration: none;

  &:after {
    content: '';
    display: block;
    padding-top: 6px;
    border-bottom: 2px solid ${COLORS.gray[600]};
  }

  &:hover:after {
    content: '';
    display: block;
    border-bottom: 2px solid ${COLORS.purple[700]};
  }
`;

const IconWrapper = styled.div`
  width: 42px;
  height: 42px;
  display: flex;
  justify-content: center;
  align-items: center;
  border: 2px solid ${COLORS.gray[400]};
  border-radius: 50%;
  color: ${COLORS.gray[400]};
  cursor: pointer;

  &:hover {
    color: ${COLORS.purple[500]};
    border-color: ${COLORS.purple[500]};
  }
`;

export default DirectoryPicker;
