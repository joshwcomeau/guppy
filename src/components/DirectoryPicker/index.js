// @flow
export { default } from './DirectoryPicker';
