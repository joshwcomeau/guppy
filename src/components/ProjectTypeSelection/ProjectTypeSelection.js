// @flow
import React, { Fragment, PureComponent } from 'react';
import styled from 'styled-components';

import { GatsbyIcon, NextjsIcon, ReactIcon } from '../Icons';
import ButtonWithIcon from '../ButtonWithIcon';
import Spacer from '../Spacer';

import type { ProjectType } from '../../types';
type Props = {
  projectType: ?ProjectType,
  onProjectTypeSelect: (projectType: ProjectType) => void,
};

class ProjectTypeSelection extends PureComponent<Props> {
  select = (ev: SyntheticEvent<*>, projectType: ProjectType) => {
    ev.preventDefault();
    this.props.onProjectTypeSelect(projectType);
  };

  render() {
    const { projectType } = this.props;
    return (
      <ProjectTypeTogglesWrapper>
        {mapProjectTypeToComponent.map((curProjectType, index) => (
          <Fragment key={index}>
            <ButtonWithIcon
              showStroke={projectType === curProjectType.type}
              icon={curProjectType.Component}
              onClick={(ev: SyntheticEvent<*>) =>
                this.select(ev, curProjectType.type)
              }
            >
              {curProjectType.caption}
            </ButtonWithIcon>
            <Spacer inline size={10} />
          </Fragment>
        ))}
      </ProjectTypeTogglesWrapper>
    );
  }
}

const ProjectTypeTogglesWrapper = styled.div`
  margin-top: 8px;
  margin-left: -8px;
`;

const mapProjectTypeToComponent = [
  {
    type: 'create-react-app',
    Component: ReactIcon,
    caption: 'Vanilla React',
  },
  {
    type: 'gatsby',
    Component: GatsbyIcon,
    caption: 'Gatsby',
  },
  {
    type: 'nextjs',
    Component: NextjsIcon,
    caption: 'Next.js',
  },
];

export default ProjectTypeSelection;
