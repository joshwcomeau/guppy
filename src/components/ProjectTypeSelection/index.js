// @flow
export { default } from './ProjectTypeSelection';
