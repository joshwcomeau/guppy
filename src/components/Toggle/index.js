// @flow
export { default } from './Toggle';
