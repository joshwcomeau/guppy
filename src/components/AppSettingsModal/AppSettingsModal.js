// @flow
import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import produce from 'immer';
import fs from 'fs';
import { remote } from 'electron';

import * as actions from '../../actions';
import { getDefaultProjectPath } from '../../reducers/app-settings.reducer';
import { COLORS } from '../../constants';
import { setNested } from '../../utils';

import Modal from '../Modal';
import ModalHeader from '../ModalHeader';
import Spacer from '../Spacer';
import { FillButton } from '../Button';
import FormField from '../FormField';
import Toggle from '../Toggle';
import PixelShifter from '../PixelShifter';
import DirectoryPicker from '../DirectoryPicker';
import ProjectTypeSelection from '../ProjectTypeSelection';
import Heading from '../Heading';

import type { AppSettings } from '../../types';

const { dialog } = remote;

type Props = {
  isVisible: boolean,
  settings: AppSettings,
  hideModal: () => void,
  saveAppSettings: (newSettings: AppSettings) => void,
};

type FormFields = 'directoryPicker' | 'projectType' | 'tracking';

type State = {
  newSettings: AppSettings,
  activeField: FormFields,
};

class AppSettingsModal extends PureComponent<Props, State> {
  state = {
    newSettings: this.props.settings,
    activeField: 'directoryPicker',
  };

  saveSettings = (ev: SyntheticEvent<*>) => {
    const { saveAppSettings } = this.props;
    const { newSettings } = this.state;
    ev.preventDefault();

    // Check if project folder exists
    // todo: Later check not exact path as one directory level is allowed to create. For now it has to exist completely.
    if (!fs.existsSync(newSettings.general.defaultProjectPath)) {
      // Path not found. Show dialog and bail early
      return dialog.showErrorBox(
        "Path doesn't exist.",
        'Please check your default project path or use the directory picker to select the path.'
      );
    }

    saveAppSettings(newSettings);
  };

  // Update setting by key
  // Notice: Key will be nested e.g. 'general.defaultProjectPath' that's why we're using utility function setNested
  updateSetting = (key: string, value: any) => {
    this.setState(state =>
      // produce could be used with currying here - so no state passed to produce. But Flow couldn't detect the right return type.
      produce(state, draftState => {
        setNested(draftState.newSettings, key, value);
      })
    );
  };

  setActive = (field: FormFields) => {
    this.setState({
      activeField: field,
    });
  };

  render() {
    const { hideModal, isVisible } = this.props;
    const { newSettings, activeField } = this.state;

    return (
      <Modal isVisible={isVisible} onDismiss={hideModal}>
        <ModalHeader title="Preferences" />

        <MainContent>
          <form onSubmit={this.saveSettings}>
            {/* // todo: Refactor to map over settings so this renders dynamically -
                         needs additional info on state e.g. component for rendering + props
                         -> For now it's OK to hard code as we're only having three
                            settings 
            */}
            <SectionTitle>General</SectionTitle>
            <PixelShifter x={5} reason="Slightly intend in section">
              <FormField label="Default Project Path" focusOnClick={false}>
                <Spacer size={5} />
                <DirectoryPicker
                  onFocus={() => this.setActive('directoryPicker')}
                  inputEditable={true}
                  path={newSettings.general.defaultProjectPath}
                  onSelect={path =>
                    this.updateSetting('general.defaultProjectPath', path)
                  }
                  isFocused={activeField === 'directoryPicker'}
                />
              </FormField>
              <ProjectTypeSelection
                label="Default Project Type"
                activeField={activeField}
                projectType={newSettings.general.defaultProjectType}
                onSelect={projectType => {
                  this.setActive('projectType');
                  this.updateSetting('general.defaultProjectType', projectType);
                }}
              />
            </PixelShifter>
            <SectionTitle>Privacy</SectionTitle>
            <PixelShifter x={5} reason="Slightly intend in section">
              <FormField
                label="Enable anonymous usage tracking"
                focusOnClick={false}
                isFocused={activeField === 'tracking'}
              >
                <Toggle
                  isToggled={newSettings.privacy.enableUsageTracking}
                  onToggle={value => {
                    this.setActive('tracking');
                    this.updateSetting('privacy.enableUsageTracking', value);
                  }}
                />
              </FormField>
            </PixelShifter>
            <Spacer size={10} />
            <Actions>
              <FillButton
                size="large"
                colors={[COLORS.green[700], COLORS.lightGreen[500]]}
              >
                Save
              </FillButton>
            </Actions>
          </form>
        </MainContent>
      </Modal>
    );
  }
}

const MainContent = styled.section`
  padding: 25px;
`;

const SectionTitle = styled(Heading).attrs({
  size: 'small',
})`
  padding-bottom: 20px;
`;

const Actions = styled.div`
  text-align: center;
  padding-bottom: 16px;
`;

const mapStateToProps = state => {
  return {
    projectHomePath: getDefaultProjectPath(state),
    settings: state.appSettings,
    isVisible: state.modal === 'app-settings',
  };
};

export default connect(
  mapStateToProps,
  {
    hideModal: actions.hideModal,
    saveAppSettings: actions.saveAppSettingsStart,
  }
)(AppSettingsModal);
