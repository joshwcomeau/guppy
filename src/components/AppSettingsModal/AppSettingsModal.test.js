import React from 'react';
import { remote } from 'electron'; // mocked
import fs from 'fs';
import { mount, shallow } from 'enzyme';

import { AppSettingsModal } from './AppSettingsModal';
import Toggle from '../Toggle';
import DirectoryPicker from '../DirectoryPicker';
import ProjectTypeSelection from '../ProjectTypeSelection';

const { dialog } = remote;

describe('AppSettingsModal component', () => {
  let wrapper;
  let instance;
  let mockedHideModal;
  let mockedSaveAppSettings;
  let mockFs;

  const DEFAULT_SETTINGS = {
    general: {
      defaultProjectPath: 'user/guppy',
      defaultProjectType: 'create-react-app',
    },
    privacy: {
      enableUsageTracking: true,
    },
  };

  beforeEach(() => {
    mockFs = jest.spyOn(fs, 'existsSync').mockImplementation(() => true);
    mockedHideModal = jest.fn();
    mockedSaveAppSettings = jest.fn();
    wrapper = shallow(
      <AppSettingsModal
        isVisible={true}
        settings={DEFAULT_SETTINGS}
        hideModal={mockedHideModal}
        saveAppSettings={mockedSaveAppSettings}
      />
    );
    instance = wrapper.instance();
  });

  afterEach(() => {
    mockFs.mockRestore();
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should show dialog if path not exists', () => {
    jest.spyOn(fs, 'existsSync').mockImplementation(() => false);
    instance.saveSettings({ preventDefault: jest.fn() });

    expect(dialog.showErrorBox).toHaveBeenCalledWith(
      "Path doesn't exist.",
      'Please check your default project path or use the directory picker to select the path.'
    );
  });

  it('should update & save settings', () => {
    instance.updateSetting('general.defaultProjectType', 'gatsby');

    instance.saveSettings({ preventDefault: jest.fn() });
    expect(mockedSaveAppSettings).toHaveBeenCalledWith({
      ...DEFAULT_SETTINGS,
      general: {
        ...DEFAULT_SETTINGS.general,
        defaultProjectType: 'gatsby',
      },
    });
  });

  it('should update path setting', () => {
    const picker = wrapper.find(DirectoryPicker);
    const pickedPath = '/new/path';
    instance.updateSetting = jest.fn();
    picker.prop('onSelect')(pickedPath);

    expect(instance.updateSetting).toHaveBeenCalledWith(
      'general.defaultProjectPath',
      pickedPath
    );
  });

  describe('activeField update', () => {
    beforeEach(() => {
      // Mounted to DOM as we'd like to trigger events
      wrapper = mount(
        <AppSettingsModal
          isVisible={true}
          settings={DEFAULT_SETTINGS}
          hideModal={mockedHideModal}
          saveAppSettings={mockedSaveAppSettings}
        />
      );
      instance = wrapper.instance();
    });

    it('should set "tracking" on privacy toggle', () => {
      const toggle = wrapper.find(Toggle);
      toggle.simulate('click');
      expect(wrapper.state('activeField')).toEqual('tracking');
    });

    it('should set "directory" on focus', () => {
      const picker = wrapper.find(DirectoryPicker);
      wrapper.setState({ activeField: 'tracking' }); // Preset to tracking
      expect(wrapper.state('activeField')).toEqual('tracking');
      picker.prop('onFocus')();
      expect(wrapper.state('activeField')).toEqual('directoryPicker');
    });

    it('should set "projectType" on select', () => {
      const selection = wrapper.find(ProjectTypeSelection);
      instance.updateSetting = jest.fn();
      selection.prop('onProjectTypeSelect')('gatsby');

      expect(wrapper.state('activeField')).toEqual('projectType');
      expect(instance.updateSetting).toHaveBeenCalledWith(
        'general.defaultProjectType',
        'gatsby'
      );
    });
  });
});
