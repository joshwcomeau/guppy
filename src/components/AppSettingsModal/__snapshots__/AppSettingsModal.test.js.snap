// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppSettingsModal component should render 1`] = `
<Modal
  isVisible={true}
  onDismiss={[MockFunction]}
  width={750}
>
  <ModalHeader
    theme="standard"
    title="Preferences"
  />
  <styled.section>
    <form
      onSubmit={[Function]}
    >
      <Styled(Heading)>
        General
      </Styled(Heading)>
      <PixelShifter
        reason="Slightly intend in section"
        x={5}
      >
        <FormField
          focusOnClick={false}
          label="Default Project Path"
          spacing={30}
        >
          <styled.div
            size={5}
          />
          <DirectoryPicker
            inputEditable={true}
            isFocused={true}
            onFocus={[Function]}
            onSelect={[Function]}
            path="user/guppy"
            pathToSelectMessage="Select the directory"
          />
        </FormField>
        <FormField
          isFocused={false}
          label="Default Project Type"
          spacing={30}
        >
          <ProjectTypeSelection
            onProjectTypeSelect={[Function]}
            projectType="create-react-app"
          />
        </FormField>
      </PixelShifter>
      <Styled(Heading)>
        Privacy
      </Styled(Heading)>
      <PixelShifter
        reason="Slightly intend in section"
        x={5}
      >
        <FormField
          focusOnClick={false}
          isFocused={false}
          label="Enable anonymous usage tracking"
          spacing={30}
        >
          <Toggle
            isToggled={true}
            onToggle={[Function]}
            padding={2}
            size={32}
          />
        </FormField>
      </PixelShifter>
      <styled.div
        size={10}
      />
      <styled.div>
        <FillButton
          colors={
            Array [
              "#00C853",
              "#69db0d",
            ]
          }
          size="large"
          textColor="#FFF"
        >
          Save
        </FillButton>
      </styled.div>
    </form>
  </styled.section>
</Modal>
`;
